lib/filewatch/fingerprinter.rb

```diff
@@ -0,0 +1,130 @@
+require 'filewatch/boot_setup' unless defined?(FileWatch)
+
+module FileWatch
+  class Fingerprinter
+    attr_reader :data, :fnv, :fingerprint, :offset, :size
+
+    def initialize(path, offset)
+      @path = path
+      @offset = offset
+      @size = FP_BYTE_SIZE
+    end
+
+    def match_any?(keys)
+      if @fnv.nil?
+        # STDERR.puts "Fingerprinter: fnv nil for path - #{path}"
+        return nil
+      end
+      if @data.nil?
+        # STDERR.puts "Fingerprinter: data nil for path - #{path}"
+        return nil
+      end
+      keys.each do |k|
+        fp, off, len = *k
+        # STDERR.puts "Fingerprinter: any_match? - #{fp}:#{off}:#{len}"
+        if off != @offset
+          # STDERR.puts "Fingerprinter: any_match? - wrong offset, mine is: #{@offset}"
+          next
+        end
+        possible = @fnv.fnv1a64(len)
+        if fp == possible
+          @fingerprint, @size = fp, len
+          return k
+        end
+      end
+      # STDERR.puts "Fingerprinter: no match - #{keys.inspect}"
+      nil
+    end
+
+    def add_size(size)
+      if @fnv
+        fnv_fingerprint(size)
+      end
+      self
+    end
+
+    def add_data(data)
+      @data = data
+      set_fnv
+      self
+    end
+
+    def read_path
+      begin
+        file = FileOpener.open(@path)
+        set_fingerprint(file)
+      ensure
+        file.close
+      end
+      self
+    end
+
+    def read_file(file)
+      set_fingerprint(file)
+      self
+    end
```

Comment:
Chatted w/ @guyboertje over zoom about this. I agree it will have a side effect of moving the read cursor of the file. We could document that the caller (of read_file) is responsible for this safety, but we can do this instead, I think, to not have any side effects:

```ruby
def read_file(file)
  dup = file.dup
  set_fingerprint(dup)
  self
ensure
  # close dup, but it could be nil if `file.dup` fails (out of files, etc)
  dup.close if dup
end
```

This will copy the file descriptor, use it for the fingerprinting, and leave the original file alone. No side effects woo!